src/session.ts

```diff
@@ -232,12 +245,29 @@ export class Session {
             if (allowModify) {
                 request = response.request.clone()
             }
-            // If signatures were returned, append them
-            if (response.signatures) {
+
+            // If signatures were returned, record them in the response.
+            if (response.signatures?.length) {
+                // Merge new signatures alongside existing signatures into the TransactResult.
                 result.signatures = [...result.signatures, ...response.signatures]
+
+                // Recover the keys used to generate the signatures at the time of the request.
+                const recoveredKeys = response.signatures.map((signature) => {
+                    const requestTransaction = request.getRawTransaction()
+                    const requestDigest = requestTransaction.signingDigest(this.chain.id)
+                    return signature.recoverDigest(requestDigest)
+                })
+
+                // Merge newly discovered keys into the TransactResult.
+                result.keys = [...result.keys, ...recoveredKeys]
             }
         }
 
+        // Validate all the signatures returned by the plugins against the current request
+        if (willValidatePluginSignatures) {
+            this.validateBeforeSignSignatures(context, request, result)
```

Comment:
Once the transaction is finalized (after all `beforeSign` hooks), and before the wallet signs the transaction, validate all of the signatures provided against the transaction the user is about to sign.

This will throw an error if any of the signatures provided by the plugins are invalid (since the transaction would fail upon broadcast anyways).